warehouse/age-off-utils/src/main/java/datawave/age/off/util/AgeOffFileGenerator.java

```diff
@@ -0,0 +1,69 @@
+package datawave.age.off.util;
+
+import java.io.IOException;
+import java.io.Writer;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Using an AgeOffFileConfiguration, writes an xml age off file containing an optional parent reference and a set of ordered rules.
+ */
+public class AgeOffFileGenerator {
+    private static final Logger log = LoggerFactory.getLogger(AgeOffFileGenerator.class);
+
+    private final AgeOffFileConfiguration configuration;
+    private final String indent;
+    private final Writer writer;
+
+    public AgeOffFileGenerator(AgeOffFileConfiguration configuration) {
+        this.configuration = configuration;
+        this.indent = this.configuration.getIndentation();
+        this.writer = this.configuration.getWriter();
+    }
+
+    /**
+     * Writes the file to the writer.
+     *
+     * @throws IOException
+     *             i/o exception with writer
+     */
+    public void format() throws IOException {
+        openConfigurationElement();
+        writerParentElement();
+        writerRules();
+        closeConfiguration();
+    }
+
+    private void closeConfiguration() throws IOException {
+        this.writer.write("</ageoffConfiguration>\n");
+    }
+
+    private void writerParentElement() throws IOException {
+        String parentFileName = this.configuration.getParentFileName();
+
+        if (null != parentFileName) {
+            log.debug("Writing parent file name: {}", parentFileName);
+            this.writer.write(this.indent + "<parent>" + parentFileName + "</parent>\n");
+        }
+    }
+
+    private void writerRules() throws IOException {
+        this.writer.write(this.indent + "<rules>\n");
+
+        for (AgeOffRuleConfiguration ruleConfiguration : this.configuration.getRuleConfigurations()) {
+            writeRule(ruleConfiguration);
+        }
+
+        this.writer.write(this.indent + "</rules>\n");
+    }
+
+    private void writeRule(AgeOffRuleConfiguration ruleConfiguration) throws IOException {
+        log.debug("formatting ruleConfiguration {}", ruleConfiguration.getRuleLabel());
+        new AgeOffRuleFormatter(ruleConfiguration).format();
```

Comment:
The writer object is sorta creeping across all of the Configuration and Formatter objects. I like the division between Conf vs Formatter where the Conf holds all the state we need to write into the XML and the Formatter knows how to convert a Conf to XML. But the writer creeping into the Conf objects makes it a bit hard to keep track of. 

What do you think about eliminating the Writer from any Conf objecs and requiring the Writer to be passed into any Formatter calls? If we need to keep track of the indent level, we could make like an XMLIndentingWriter like you have already but maybe open up an "indent()" and "outdent()" API.


```
writer = new IndentingXMLWriter(rawWriter);

// in AgeOffFileGenerator
for ruleConf in rulesConfs
  writer.indent();
  new AgeOffRuleFormatter(ruleConf).format(writer);
  writer.outdent();

// in AgeOffRuleFormatter
new AgeOffCsvToMatchPatternFormatter(matchConf).format(writer);

```

Reply:
I removed the reader and writer from the Configuration objects.  I think that was good feedback.  The writer is now being passed to the format method which seems cleaner.  I stuck with the delegating indenting writer approach instead of changing it to do the indent/outdent.  Hopefully it looks ok this way now that the writer is passed to format.